pupil-api/src/services/pupil-authentication.service.ts

```diff
@@ -0,0 +1,89 @@
+'use strict'
+
+import * as azureStorage from 'azure-storage'
+import * as bluebird from 'bluebird'
+import { clone } from 'ramda'
+import * as winston from 'winston'
+
+let azureTableService: any
+const authTable = 'preparedCheck'
+
+export const pupilAuthenticationService = {
+  authenticate: async function authenticate (pupilPin: string, schoolPin: string, tableService?: any): Promise<object> {
+    // set tableService to azureTableService if not provided, but only instantiate it once.
+    if (!tableService) {
+      if (!azureTableService) {
+        azureTableService = azureStorage.createTableService()
+        bluebird.promisifyAll(azureTableService, {
+          promisifier: (originalFunction) => function (...args) {
+            return new Promise((resolve, reject) => {
+              try {
+                originalFunction.call(this, ...args, (error, result, response) => {
+                  if (error) {
+                    return reject(error)
+                  }
+                  resolve({ result, response })
+                })
+              } catch (error) {
+                reject(error)
+              }
+            })
+          }
+        })
+      }
+      tableService = azureTableService
+    }
+
+    const { result } = await tableService.retrieveEntityAsync(authTable, schoolPin, pupilPin)
+
+    // Prepare the pupil data for use by the SPA
+    const data = this.preparePupilData(result)
+
+    // Mark the data as having been collected
+    try {
+      await this.markAsCollected(tableService, clone(result))
+    } catch (error) {
+      winston.error(error.message)
+      throw error
```

Comment:
A quick note for `markAsCollected`, there is no other way for this to mark the result as corrected and not return the data (which would lead to login failing but being marked as collected and the pupil won’t be able to login anymore), right? Seems like that to me 👍 or we might need to retry somehow?

Reply:
this piece of code may well be removed Dan and passed off to a queue message and corresponding function bound to the queue.  I would prefer the login API to just serve data, nothing more.  However, this was part of the original design so completely understand why Jon has put it in.  I will add the retry policy to the upcoming card that has the refinements in it.

Reply:
This behaviour has now been relaxed - repeat logins are now allowed.